### PR TITLE
async/await support

### PR DESCRIPTION
**Pull request by fabianfett**
`async`/`await` is awesome! Let's support this in Lambda.

Modifications:

- Add an `AsyncLambdaHandler`. Will be renamed to `LambdaHandler` as soon as we drop the current callback based `LambdaHandler`.
- The default way to use an `AsyncLambdaHandler` is to use `@main` to execute it. Don't use `Lambda.run` for it. We wan't to remove `Lambda.run` for 1.0.

### CODE REVIEW

**Comment by pokryfka** [comment]
Node.js developers will feel at home :-)

But we would need `swift-no` support it to be useful inside Lambda handler implementations, no?

**Comment by fabianfett** [comment]
This might be interesting to us. Potentially replacing the `LambdaHandlerFactory`:

https://github.com/apple/swift/pull/35398

File: Sources/AWSLambdaRuntime/Lambda+Codable.swift
Comment (weissi) [inline Sources/AWSLambdaRuntime/Lambda+Codable.swift:99]: I would advise against typealiases for closures, they give you really bad completions and it's unclear to the user what they're supposed to put there (because they'll just see `CodableVoidAsyncClosure<In>`).

File: Sources/AWSLambdaRuntime/Lambda+Codable.swift
Comment (weissi) [inline Sources/AWSLambdaRuntime/Lambda+Codable.swift:86]: context should be the rightmost argument

File: Sources/AWSLambdaRuntime/Lambda+Codable.swift
Comment (weissi) [inline Sources/AWSLambdaRuntime/Lambda+Codable.swift:107]: overloading on closure arguments is probably not a good idea

File: Sources/AWSLambdaRuntimeCore/Lambda.swift
Comment (weissi) [inline Sources/AWSLambdaRuntimeCore/Lambda.swift:75]:
```suggestion
    let promise = context.eventLoop.makePromise(of: ...)
    promise.completeWithAsync {
        try await runThing()
    }
    return promise.futureResult
}
```

File: Sources/AWSLambdaRuntimeCore/LambdaHandler.swift
Comment (weissi) [inline Sources/AWSLambdaRuntimeCore/LambdaHandler.swift:109]: I don't think we need any `@asyncHandler`s in here.

File: Sources/AWSLambdaRuntimeCore/LambdaHandler.swift
Comment (weissi) [inline Sources/AWSLambdaRuntimeCore/LambdaHandler.swift:90]: hmm, an `AsyncLambdaHandler` isn't running the async code on the `EventLoop` at the moment (that's impossible without custom executors), is this proto conformance still right?

File: Sources/AWSLambdaRuntimeCore/LambdaHandler.swift
Comment (weissi) [inline Sources/AWSLambdaRuntimeCore/LambdaHandler.swift:104]: this should use `promise.completeWithAsync` I think

**Review by weissi — CHANGES_REQUESTED** [review]
Awesome, this looks great. Left a few comments, mostly minor around overloading / `@asyncHandler` etc

File: Sources/AWSLambdaRuntime/Lambda+Codable.swift
Comment (fabianfett) [reply Sources/AWSLambdaRuntime/Lambda+Codable.swift:99]: I support your argument. 👍 
However if we don't use a `typealias` here, we should later adjust all other the callback closures as well. @tomerd wdyt?

File: Sources/AWSLambdaRuntimeCore/LambdaHandler.swift
Comment (fabianfett) [reply Sources/AWSLambdaRuntimeCore/LambdaHandler.swift:90]: Yes it is. Those protocols build up on top of each other:

File: Package.swift
Comment (weissi) [inline Package.swift:61]: you can always depend on it, `_NIOConcurrency` just doesn't export anything if A/A isn't on. But up to you

File: Sources/AWSLambdaRuntime/Lambda+Codable.swift
Comment (tomerd) [inline Sources/AWSLambdaRuntime/Lambda+Codable.swift:83]: this is good stuff. should we also mark the closure ones deprecated? I would argue we should.

File: Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift
Comment (tomerd) [inline Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift:20]: `// Mark : - ...`

File: Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift
Comment (tomerd) [inline Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift:145]: `// Mark : - ...`

File: Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift
Comment (tomerd) [inline Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift:207]: `// Mark : - ...`

File: Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift
Comment (tomerd) [inline Tests/AWSLambdaRuntimeCoreTests/LambdaHandlerTest.swift:84]:
```suggestion
    // MARK: - AsyncLambdaHandler
```

File: Package.swift
Comment (fabianfett) [inline Package.swift:18]: We need to wait for the next NIO release before we can merge. @weissi ;)

**Comment by fabianfett** [comment]
@swift-server-bot test this please

**Review by weissi — APPROVED** [review]
nice one! lgtm!

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
awesome!

**Pull request merged**
